packages/typegpu-react/src/use-uniform-value.ts

```diff
@@ -1,13 +1,55 @@
 import type * as d from 'typegpu/data';
+import { useRoot } from './root-context.tsx';
+import { useEffect, useMemo, useState } from 'react';
+import type { ValidateUniformSchema } from 'typegpu';
 
 interface UniformValue<TSchema, TValue extends d.Infer<TSchema>> {
   schema: TSchema;
-  value: TValue;
+  value: TValue | undefined;
+  readonly $: d.InferGPU<TSchema>;
 }
 
-export function useUniformValue<TSchema, TValue extends d.Infer<TSchema>>(
-  schema: d.AnyWgslData,
+export function useUniformValue<
+  TSchema extends d.AnyWgslData,
+  TValue extends d.Infer<TSchema>,
+>(
+  schema: ValidateUniformSchema<TSchema>,
   initialValue?: TValue | undefined,
 ): UniformValue<TSchema, TValue> {
-  // TODO: Implement
+  const root = useRoot();
+
+  const [uniformBuffer] = useState(() => {
+    return root.createUniform(
+      schema,
+      initialValue,
+    );
+  });
+
+  useEffect(() => {
+    return () => {
+      uniformBuffer.buffer.destroy();
+    };
+  }, [uniformBuffer]);
+
+  // biome-ignore lint/correctness/useExhaustiveDependencies: This value needs to be stable
+  const uniformValue = useMemo(() => {
+    let currentValue = initialValue;
+    return {
+      schema,
+      get value() {
+        return currentValue;
+      },
+      set value(newValue: TValue | undefined) {
+        currentValue = newValue;
+        if (newValue !== undefined) {
+          uniformBuffer.write(newValue);
+        }
+      },
```

Comment:
Because of the previous comment, we don't have to worry about undefined.
```suggestion
      get value() {
        return currentValue;
      },
      set value(newValue: TValue) {
        currentValue = newValue;
        uniformBuffer.write(newValue);
      },
```